File: web/models/offline-message.ts
import type { AgentBrief, ChannelBrief, Conversation, GroupBrief, Message, VisitorBrief } from '@/models/conversation'

export type OfflineMessageStatus = 'pending' | 'converted'

export type OfflineMessageEntry = Omit<Message, 'conversation_id' | 'conversation_public_id'> & {
  offline_message_id: number
}

export type OfflineMessage = {
  id: number
  public_id: string
  tenant_id: number
  status: OfflineMessageStatus
  visitor: VisitorBrief | null
  channel: ChannelBrief | null
  target_group: GroupBrief | null
  conversation: Conversation | null
  visitor_external_id: string
  visitor_name: string | null
  handled_by_id: number | null
  handled_at: string | null
  last_message_at: string | null
  last_message_preview: string | null
  message_count: number
  metadata?: Record<string, unknown>
  created_at: string
  updated_at: string | null
}

export type OfflineMessageDetail = OfflineMessage & {
  messages: OfflineMessageEntry[]
  has_more_messages: boolean
  can_assign_self?: boolean
  can_assign_other?: boolean
}

export type OfflineMessageListResponse = {
  items: OfflineMessage[]
  has_more: boolean
  total?: number | null
}

export type OfflineMessageCountResponse = {
  total: number
}

export type OfflineMessageConvertResponse = {
  offline_message: OfflineMessage
  conversation: Conversation
  messages: Message[]
  assigned_to_current_user?: boolean
  assigned_agent?: AgentBrief | null
}

File: server/app/schemas/offline_message.py
"""
Offline message schemas.
"""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.conversation import AgentBrief, ChannelBrief, ConversationResponse, GroupBrief, VisitorBrief
from app.schemas.message import MessageResponse, PublicMessageResponse


class OfflineMessageCreateRequest(BaseModel):
    visitor_name: str | None = Field(default=None, max_length=128)
    metadata: dict | None = None
    system: str | None = Field(default=None, max_length=64)
    browser: str | None = Field(default=None, max_length=128)


class OfflineMessageSendRequest(BaseModel):
    content_type: Literal["text", "image", "file"] = "text"
    content: str = Field(..., max_length=20000)
    system: str | None = Field(default=None, max_length=64)
    browser: str | None = Field(default=None, max_length=128)


class OfflineMessageEntryResponse(BaseModel):
    id: int
    offline_message_id: int
    sender_type: str
    sender_id: int | None = None
    sender_name: str | None = None
    sender_avatar: str | None = None
    content_type: str
    content: str
    metadata: dict = Field(default_factory=dict)
    created_at: datetime


class OfflineMessageBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    public_id: str
    tenant_id: int
    status: Literal["pending", "converted"]
    visitor: VisitorBrief | None = None
    channel: ChannelBrief | None = None
    target_group: GroupBrief | None = None
    conversation: ConversationResponse | None = None
    visitor_external_id: str
    visitor_name: str | None = None
    handled_by_id: int | None = None
    handled_at: datetime | None = None
    last_message_at: datetime | None = None
    last_message_preview: str | None = None
    message_count: int = 0
    metadata: dict = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime | None = None


class OfflineMessageDetail(OfflineMessageBrief):
    messages: list[OfflineMessageEntryResponse] = Field(default_factory=list)
    has_more_messages: bool = False
    can_assign_self: bool = False
    can_assign_other: bool = False


class OfflineMessageListResponse(BaseModel):
    items: list[OfflineMessageBrief]
    has_more: bool
    total: int | None = None


class OfflineMessageCountResponse(BaseModel):
    total: int


class PublicOfflineMessageResponse(BaseModel):
    offline_message_public_id: str
    status: Literal["pending", "converted"]
    messages: list[PublicMessageResponse] = Field(default_factory=list)
    has_more: bool = False
    conversation_public_id: str | None = None


class OfflineMessageAssignSelfRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=200)


class OfflineMessageAssignRequest(OfflineMessageAssignSelfRequest):
    agent_id: int


class OfflineMessageConvertResponse(BaseModel):
    offline_message: OfflineMessageBrief
    conversation: ConversationResponse
    messages: list[MessageResponse] = Field(default_factory=list)
    assigned_to_current_user: bool = True
    assigned_agent: AgentBrief | None = None
